### ISSUE TITLE
appsody list fails when no connectivity (wifi turned off).

### ISSUE THREAD

**Issue by bradtopol** [appsody/appsody#265]
**Describe the bug**
I have appsody installed an running correctly on my mac using docker desktop and its Kubernetes.  I plan on running a simple appsody demo at an upcoming summit doing some basic appsody list, appsody init, appsody run etc. I decided to try this with wifi turned off just in case the Summit does not have working wifi. When I try this I get the following:
Brads-MBP-2:demo4 bradtopol$ appsody list
[Warning] Unable to check the most recent version of Appsody in GitHub.... continuing.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x1470a9b]

goroutine 1 [running]:
github.com/appsody/appsody/cmd.getLatestVersion(0x0, 0x1000000014c8e40)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/utils.go:970 +0x5b
github.com/appsody/appsody/cmd.doVersionCheck(0xc000190000, 0xc8, 0x2c8, 0xc000020580, 0x1d, 0xc000020780, 0x1d, 0xc000154570, 0x27)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/utils.go:977 +0x37
github.com/appsody/appsody/cmd.checkTime()
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/utils.go:1014 +0x2ec
github.com/appsody/appsody/cmd.setupConfig(0x1526ac0, 0xc00002a500)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/root.go:80 +0x5c
github.com/appsody/appsody/cmd.glob..func11(0x19c0b00, 0x19e7c30, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/list.go:30 +0x51
github.com/appsody/appsody/vendor/github.com/spf13/cobra.(*Command).execute(0x19c0b00, 0x19e7c30, 0x0, 0x0, 0x19c0b00, 0x19e7c30)
	/home/travis/gopath/src/github.com/appsody/appsody/vendor/github.com/spf13/cobra/command.go:762 +0x465
github.com/appsody/appsody/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19bf800, 0x147136e, 0x19bf800, 0xc00011bf88)
	/home/travis/gopath/src/github.com/appsody/appsody/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
github.com/appsody/appsody/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/github.com/appsody/appsody/vendor/github.com/spf13/cobra/command.go:800
github.com/appsody/appsody/cmd.Execute(0x1621f24, 0x5)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/root.go:159 +0x57
main.main()
	/home/travis/gopath/src/github.com/appsody/appsody/main.go:26 +0x39
 

**To Reproduce**
1. turn off wifi
2. type appsody list
3. 
4. See error:
Brads-MBP-2:demo4 bradtopol$ appsody list
[Warning] Unable to check the most recent version of Appsody in GitHub.... continuing.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x1470a9b]

goroutine 1 [running]:
github.com/appsody/appsody/cmd.getLatestVersion(0x0, 0x1000000014c8e40)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/utils.go:970 +0x5b
github.com/appsody/appsody/cmd.doVersionCheck(0xc000190000, 0xc8, 0x2c8, 0xc000020580, 0x1d, 0xc000020780, 0x1d, 0xc000154570, 0x27)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/utils.go:977 +0x37
github.com/appsody/appsody/cmd.checkTime()
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/utils.go:1014 +0x2ec
github.com/appsody/appsody/cmd.setupConfig(0x1526ac0, 0xc00002a500)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/root.go:80 +0x5c
github.com/appsody/appsody/cmd.glob..func11(0x19c0b00, 0x19e7c30, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/list.go:30 +0x51
github.com/appsody/appsody/vendor/github.com/spf13/cobra.(*Command).execute(0x19c0b00, 0x19e7c30, 0x0, 0x0, 0x19c0b00, 0x19e7c30)
	/home/travis/gopath/src/github.com/appsody/appsody/vendor/github.com/spf13/cobra/command.go:762 +0x465
github.com/appsody/appsody/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19bf800, 0x147136e, 0x19bf800, 0xc00011bf88)
	/home/travis/gopath/src/github.com/appsody/appsody/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
github.com/appsody/appsody/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/github.com/appsody/appsody/vendor/github.com/spf13/cobra/command.go:800
github.com/appsody/appsody/cmd.Execute(0x1621f24, 0x5)
	/home/travis/gopath/src/github.com/appsody/appsody/cmd/root.go:159 +0x57
main.main()
	/home/travis/gopath/src/github.com/appsody/appsody/main.go:26 +0x39

**Expected behavior**

**Actual behaviour**
What is the actual behaviour.

**Environment Details (please complete the following information):**
 - OS: [e.g. iOS] Mac
 - Browser [e.g. chrome, safari]chrome

**If applicable please specify:**
 - CLI version:
 - Stack you are using:

**Screenshots**

**Additional context**

**Comment by Kamran64** [comment]
The latest release of Appsody (0.4.1) added a feature to check if the newest version of the CLI is available by checking GitHub for the latest version. It's possible there's no safeguarding put in place in case the user is not connected to the Internet.

**Comment by skoh7645** [comment]
Thank you @bradtopol for raising this issue. I have put in a PR to fix this bug https://github.com/appsody/appsody/pull/266

**Comment by Kamran64** [comment]
The PR has been merged to fix that bug. However Appsody requires a working Internet connection to run commands (as some need to access Dockerhub)

**Comment by bradtopol** [comment]
@kamran I thought if I have already pulled down base images ahead of time that they would be cached locally. Is that not the case?

**Comment by Kamran64** [comment]
Hi @bradtopol - Did some testing and it seems like it depends on which stack you're using. For example `nodejs` requires a connection to npm registry which requires WiFi. To answer your question though, images _should_ be cached locally as I just tried with Python-Flask as well and that works. Which stack were you using?

@neeraj-laad might be able to give a more concise answer. Apologies for the delay in response.

**Comment by neeraj-laad** [comment]
@bradtopol there are a number of things here:

- appsody cli checks if you have the latest cli version. This is the problem u were facing. We have resolved this by updating the cli to skip this if there is no network.

- `appsody list` command reads index from latest appsody stack release. This needs network connection, if you do not have a network connection, we give a specific message indicating which repo we could not read. 

- `appsody run/debug/test` will use local images if we can't pull the latest image from dockerhub. However, the stacks themselves might need network to download package dependencies etc.  

In summary, appsody cli now does not break unexpectedly with a stack trace etc. when network is not available, and we provide a meaningful error message in such situations. 

However, for developers to properly work with appsody a network connection is still vital. 

I propose we close this issue and create a new issue if there are use-case where we need a complete offline mode for appsody.

**Comment by sympatheticmoose** [comment]
@neeraj-laad do we state anywhere in the docs that a network connection is required?

**Comment by neeraj-laad** [comment]
Good point. I've created an issue to deal with appropriate documentation. https://github.com/appsody/website/issues/284

**Comment by erpenbach** [comment]
Does this mean that we do not support an airgap scenario?

**Comment by kylegc** [comment]
At this time, I would not claim "support" for airgap, but it should still be possible with the right setup. And there's a lot of setup required. You'd have to have "local" repositories for all your dependencies, such as artifactory for maven, npm, etc. A local docker repository for the stack images, and probably modify the stacks to point to the local dependencies. Finally you'd need a store for the stack template tars and the repo index. Then you'd have to set your CLIs (at least appsody and docker) to point to the local URLs etc. I'm sure I missed some steps.

**Comment by erpenbach** [comment]
Ok then I assume @neeraj-laad comment on appropriate documentation covers the air gap setup and not just a limitation statement.
